diff -urN crossdev-20120721.old/wrappers/site/linux crossdev-20120721/wrappers/site/linux
--- crossdev-20120721.old/wrappers/site/linux	2012-09-17 00:01:41.391848698 +0200
+++ crossdev-20120721/wrappers/site/linux	2012-09-17 00:09:45.970436124 +0200
@@ -49,3 +49,8 @@
 screen_cv_sys_sockets_nofs=${screen_cv_sys_sockets_nofs=no}
 screen_cv_sys_sockets_usable=${screen_cv_sys_sockets_usable=yes}
 screen_cv_sys_terminfo_used=${screen_cv_sys_terminfo_used=yes}
+
+# python
+ac_cv_have_long_long_format=${ac_cv_have_long_long_format=yes}
+ac_cv_buggy_getaddrinfo=${ac_cv_buggy_getaddrinfo=no}
+ac_cv_computed_gotos=${ac_cv_computed_gotos=yes}
