diff -urN subversion-1.7.7.new/Makefile.in subversion-1.7.7/Makefile.in
--- subversion-1.7.7.new/Makefile.in	2013-01-14 02:48:17.043010582 +0100
+++ subversion-1.7.7/Makefile.in	2013-01-14 03:05:50.074057568 +0100
@@ -355,7 +355,6 @@
 local-all: @BUILD_RULES@ @TRANSFORM_LIBTOOL_SCRIPTS@
 
 transform-libtool-scripts: @BUILD_RULES@
-	@$(SHELL) $(top_srcdir)/build/transform_libtool_scripts.sh
 
 locale-gnu-pot:
 	cd $(abs_srcdir) && XGETTEXT="$(XGETTEXT)" MSGMERGE="$(MSGMERGE)" \
