diff -urN crossdev-20120721.old/wrappers/Makefile crossdev-20120721/wrappers/Makefile
--- crossdev-20120721.old/wrappers/Makefile	2012-09-17 00:01:41.392848826 +0200
+++ crossdev-20120721/wrappers/Makefile	2012-09-17 00:14:15.859737678 +0200
@@ -4,7 +4,7 @@
 include ../settings.mk
 
 PORTDIR ?= $(shell portageq envvar PORTDIR)
-FNAMES = cross-emerge  cross-fix-root  cross-pkg-config  emerge-wrapper
+FNAMES = cross-emerge  cross-fix-root  cross-pkg-config  emerge-wrapper cross-emerge-ng cross-wrapper
 SITE = $(PREFIX)/share/crossdev/include/site
 TOPDIR =
 
@@ -16,8 +16,12 @@
 	cp -a etc $(DESTDIR)/$(PREFIX)/share/crossdev/
 	cp -a site $(DESTDIR)/$(PREFIX)/share/crossdev/include/
 	sed -i -e s@__PREFIX__@$(PREFIX)@g $(DESTDIR)/$(PREFIX)/bin/emerge-wrapper
+	sed -i -e s@__PREFIX__@$(PREFIX)@g $(DESTDIR)/$(PREFIX)/bin/cross-emerge-ng
 	sed -i -e s@__TOPDIR__@$(SITE)@g $(DESTDIR)/$(SITE)/config.site
 	mv $(DESTDIR)/$(SITE)/config.site $(DESTDIR)/$(PREFIX)/share/
 	ln -sf $(PORTDIR)/profiles/embedded $(DESTDIR)/$(PREFIX)/share/crossdev/etc/make.profile
+	
+	ln -s cross-wrapper $(DESTDIR)/$(PREFIX)/bin/cross-gcc
+	ln -s cross-wrapper $(DESTDIR)/$(PREFIX)/bin/cross-g++
 
 .PHONY: all install
